deciles_chart/core.py

```diff
@@ -72,6 +72,12 @@ def drop_zero_denominator_rows(measure_table: pandas.DataFrame) -> pandas.DataFr
 def get_deciles_table(measure_table: pandas.DataFrame) -> pandas.DataFrame:
     by = ["date"] + measure_table.attrs["group_by"][1:]
     deciles_table = measure_table.groupby(by)["value"].quantile(DECILES).reset_index()
+    # Previously, it wasn't necessary to rename "level_1", which is created by the call
+    # to .quantile, to "deciles" because this method used the name of DECILES (i.e.
+    # Series.name). However, it now is, which may be a regression in Pandas. When we
+    # downgrade Pandas (a prod dependency) to match opensafely-core/python-docker, we
+    # should revisit.
+    deciles_table = deciles_table.rename(columns={"level_1": "deciles"})
```

Comment:
I'm getting an error when I test this by hand where the column name is "level_2" rather than "level_1"
The files that I have generated recently using the measures framework have an unnamed first column which looks like an id. This results in _get_group_by returning Index(['Unnamed: 0', 'practice'], dtype='object') rather than just practice, and by to have two items.

Reply:
When you say "by hand", are you calling `deciles_chart.get_deciles_table` and passing a measures table that you have opened by calling, for example, `pandas.read_csv`? What's the error?

If a measures table generated by the measures framework has an unnamed first column that looks like an ID column, then this might be a regression in cohort-extractor. This is a common Pandas "gotcha" when `.to_csv(f_name)` is called on a dataframe; often, we want `.to_csv(f_name, index=False)`, which drops the index from the CSV.

Alternatively, and more likely, maybe cohort-extractor did this all along and there's a bug in deciles-chart-action 🙂 I'll generate a measures table and take a closer look.

Reply:
I've generated a measures table using the measures framework; here are the first five rows:

| age_band | has_copd | population |  value |       date |
| -------- | -------- | ---------- | ------ | ---------- |
| 16-29    |       13 |        123 | 0.106… | 2021-01-01 |
| 30-39    |       15 |        137 | 0.109… | 2021-01-01 |
| 40-49    |       12 |        125 | 0.096… | 2021-01-01 |
| 50-59    |       12 |        128 | 0.094… | 2021-01-01 |
| 60-69    |       11 |        112 | 0.098… | 2021-01-01 |
| ...      |      ... |        ... |    ... |        ... |

(Note that a _measures table_, as understood by deciles-chart-action, has a `date` column. cohort-extractor outputs a file for each time unit over the time range and a file for all time units over the time range. Only the latter is a measures table, according to deciles-chart-action.)

It seems as though my measures table and your measures table are different 🙂. Could you link to the study you ran to generate your measures table? (i.e. the measures table with an ID-like column) If you have a spare moment, and talking would be easier than writing, then we could have a chat.

Reply:
Sorry for the delay! The [sro-template](https://github.com/opensafely/sro-template) seems to generate measure tables with an extra column. Such as:
| | practice | event | population | value | date |
| -------- | -------- | ---------- | ------ | ---------- | ------ |
0 | 0 | 2457.0 | 5000.0 | 0.4914 | 2021-06-01 |
21| 27 | 191.0 | 359.0 | 0.5320... | 2021-06-01 |
22 | 28 | 170.0 | 326.0 | 0.5214... | 2021-06-01 |
| ...      |      ... |        ... |    ... |        ... |   ...|

I re-ran other studies such as anti-psychotics and SRO-measures that I ran in early February and they still do not have the extra column, so I do not think it is a cohort extractor regression issue. Maybe how sro-template is specifying variables and measures, but I have not been able to pin down the cause yet.

I could approve this pull request and file a separate issue for investigating the extra column and potentially handling it for decile charts?

Reply:
Ah, I think that's because sro-template includes the "expensive variable" hack:

1. Extract a cohort
2. Extract an expensive variable, which in this case is ethnicity
3. Join the cohort to the expensive variable and save a new cohort
4. Generate measures from the new cohort

My guess is that in step 3 we don't call `.to_csv(..., index=False)` when saving the data frame that contains the new cohort as a CSV. Yes, [this][1]

```python
merged_df = df.merge(ethnicity_df, how='left', on='patient_id')

merged_df.to_csv(file_path)
```

should be this

```python
merged_df = df.merge(ethnicity_df, how='left', on='patient_id')

merged_df.to_csv(file_path, index=False)
```

So, the shape of the file that is passed to `generate_measures` is slightly different in sro-template than in a study where `generate_cohort` is the parent of `generate_measures`. Consequently, I think the issue is with sro-template and not with deciles-chart-action. (Although we could tighten up the validation in deciles-chart-action to check for a column without a name.)

What do you think, @LFISHER7?

[1]: https://github.com/opensafely/sro-template/blob/master/analysis/join_ethnicity.py#L14-L16

Reply:
opensafely/sro-template#30 drops the index when writing the CSV file. I'm not sure how many studies have been created from the template, but let's not change them until changing them is necessary. Would you be happy to approve this PR now, @ccunningham101?